### ISSUE TITLE
Add CUDA losses, dropout, and remaining normalizations to unblock GPU training

### ISSUE THREAD

**Issue by xupremix** [xupremix/incin#84]
## Problem

GPU training does not work end to end, and `docs/book/src/whats_not_finished.md` names the two reasons precisely: "no accelerator has a loss, and the one that has normalization cannot run its backward."

Against the complete CPU set, CUDA is missing every loss and regularization operation:

| Category | Missing on CUDA |
|---|---|
| Loss | `bce_with_logits_loss`, `cross_entropy_loss`, `l1_loss`, `mse_loss` |
| Regularization | `dropout` |
| Normalization | `group_norm`, `instance_norm` |

`layer_norm` and `batch_norm` are advertised on CUDA as forward-only, which is what #4 addresses.

Every model in the book's Building models chapter therefore trains on CPU only. This is the smallest set of operations whose absence blocks the entire training story on the one accelerator with the widest coverage, which is why it is ahead of the larger breadth gaps.

## Proposed implementation

- Implement forward and backward CUDA executors for `mse_loss`, `l1_loss`, `cross_entropy_loss`, and `bce_with_logits_loss`.
- Compute `cross_entropy_loss` and `bce_with_logits_loss` in a numerically stable form rather than composing `log` after `softmax` or `sigmoid`, and accumulate the reduction in f32 regardless of input dtype.
- Implement `dropout` with a counter-based RNG so a mask is reproducible from seed and offset without storing it, and make the training and inference paths distinct rather than branching on a runtime rate of zero.
- Implement `group_norm` and `instance_norm` forward and backward, preserving the forward statistics for backward instead of recomputing them under different numerical conditions.
- Advertise each row as training-capable only after the conformance comparison passes.

## Acceptance criteria

- Forward values and input, weight, and bias gradients match the CPU reference within documented tolerances for all seven operations.
- Reduction modes (`mean`, `sum`, `none`) are covered where the operation has them.
- `dropout` produces the same mask for the same seed and offset, and is the identity in evaluation mode.
- A model from the Building models chapter completes a full training step on CUDA once #4 lands.
- The capability table changes only the rows these operations own.

## Proposed API and capability-row change

No new public API. These operations already exist as `Tensor` methods and as catalog rows; what changes is which backends advertise them. The user-visible effect is that code that already compiles starts running on CUDA:

```rust
use incin::prelude::*;
type Backend = DefaultBackend;   // with `cuda`, this is the CUDA backend

let logits: Tensor<s![32, 10], Backend, f32, Grad> = model.forward(x)?;
let loss = logits.cross_entropy_loss(&targets)?;   // today on CUDA: capability refusal
let grads = loss.backward()?;                      // today on CUDA: unreachable
```

The change is a capability declaration plus the executors it obligates. `crates/incin-backends/src/capability/tables.rs` drives the CUDA table from a macro over a descriptor operation list:

```rust
// crates/incin-backends/src/capability/tables.rs - proposed additions
pub static CUDA_CAPABILITIES: &[CapabilityRule] = cuda_descriptor_operations!(
    descriptor_capability_rules,
    // ...
    loss_dtypes = F32_ONLY,             // proposed: was absent
    dropout_dtypes = FLOAT_DTYPES,      // proposed: was absent
    normalization_dtypes = F32_ONLY,    // widened to cover group_norm / instance_norm
    // ...
);
```

and one executor per advertised row:

```rust
// proposed
impl Execute<op::CrossEntropyLoss> for CudaBackendImpl {
    type Output = CudaStorage;
    fn execute(&self, request: ExecutionRequest<'_, op::CrossEntropyLoss, Self>)
        -> Result<CudaStorage, BackendError>
    {
        // `request.operation` is Validated: read the derived output shape,
        // do not re-derive it.
        let out_shape = request.operation.descriptor().output_shape();
        /* fused stable log-softmax + gather + reduce */
    }
}
```

Because the CPU declaration generates a compile-time obligation that each advertised row has an `Execute` impl, advertising one of these without implementing it does not build. That is the mechanism to copy here.

## Shape and type-system invariants

| Operation | Shape rule | Numeric contract |
|---|---|---|
| `mse_loss`, `l1_loss` | `[..] , [..] -> []` or `[..]` by `Reduction` | Accumulate in f32 for reduced-precision inputs |
| `cross_entropy_loss` | `[N, C] , [N] -> []` or `[N]` | Stable form; never `log` after `softmax` |
| `bce_with_logits_loss` | `[..] , [..] -> []` or `[..]` | Stable form; never `log` after `sigmoid` |
| `dropout` | shape-preserving | Identity in eval mode; mask reproducible from `(seed, offset)` |
| `group_norm` | `[N, C, ..]` with `C % groups == 0` | Divisibility is a static fact where `C` is static |
| `instance_norm` | `[N, C, ..]` | Per-sample, per-channel statistics |

The `Reduction` types (`Mean`, `NoneReduction`) are already in the prelude, so the loss signatures are unchanged; only their backend availability moves.

## How this interoperates

- **`whats_not_finished.md` names this issue and #4 together** as the two reasons GPU training is blocked: "no accelerator has a loss, and the one that has normalization cannot run its backward." Closing both removes that entry.
- **#4** supplies LayerNorm backward. Neither issue alone produces a trainable model on CUDA.
- **#83** verifies each newly advertised row against the CPU oracle, including the `Training: yes` claim, which for losses means checking the gradient rather than only the value.
- **#90** determines whether these can run in `bf16`; landing them `f32`-only first is fine, and the capability rows should widen with that work rather than being written twice.
- **#106** supplies on-device `to_dtype`, which mixed-precision loss computation needs to avoid a host round trip.
- **`docs/capabilities.md`** regenerates from the registration. Do not hand-edit.

## Alternatives

Composing the losses from existing CUDA primitives would inherit their tape entries with no new kernels, but gives up numerical stability on exactly the operations that need it most and adds several device round trips per step.

## Additional context

Relevant code: `crates/incin-backends/src/cuda/ops/norm.rs`, `crates/incin-backends/src/cuda/ops/reduce.rs`, `crates/incin-backends/src/cuda/executor.rs`, and the CPU references under `crates/incin-backends/src/cpu/ops/`.

Pairs with #4 (CUDA LayerNorm backward). Together those two close the "GPU training is blocked" entry in `whats_not_finished.md`. Verified by #83 on the runner from #82.

**Comment by xupremix** [comment]
Measured on a GTX 1650 SUPER (sm_75), develop at 2daea9bb: the premise needs narrowing -- three of the four losses already train on CUDA through composed tape entries, and the fourth does as of this week.

| Loss | Forward | Backward | Notes |
|---|---|---|---|
| `mse_loss` | 2/3 exact | [0, -2/3, -2/3] exact | worked all along |
| `l1_loss` | 2/3 exact | [0, -1/3, -1/3] exact | needed the `unbroadcast` fix below |
| `bce_with_logits_loss` | 0.4205 | matches sigmoid-target over batch by hand | worked all along |
| `cross_entropy_loss` | 1.2144 | matches (softmax-onehot)/batch by hand | needed the gather fix below |

Two defects fixed to get there (both with red-then-green regression tests):
1. `cuda::tape::unbroadcast` handed sub-target gradients straight on; a mean-seeded scalar reached a binary launch that refuses it. It now materializes the broadcast after a compatibility check.
2. The symbolic `diff` answered -1 for d|x|/dx at 0 against CPU `Sign`, PyTorch and the hand-written kernel expression (0). The fused abs path diverged exactly where l1 losses evaluate; fixed.
3. The cross-entropy executor called the raw gather launch (no tape entry), so logits gradients were silently absent. Routed through the tracked gather.

What genuinely remains of this issue: `dropout`, `group_norm`, `instance_norm` (plus batch-norm backward, which lives outside this issue). Suggest retitling to name those three.

**Comment by xupremix** [comment]
Update: dropout trains on CUDA as well, measured just now (GTX 1650 SUPER). Forward draws once and the backward replays that exact draw: over 8 elements at p=0.5 every output was 0 or 2x input and every gradient 0 or 2.0 in the same lanes. Permanent test: `dropout_trains_through_the_replayed_mask_on_cuda`.

Revised remainder of this issue: `group_norm` and `instance_norm` have forward executors but zero test coverage of any kind on CUDA (no forward parity, no backward entries), and batch-norm backward lives in #123. Suggest retitling to the two norms.
